File: src/hideout/ask.py
from __future__ import annotations

from collections.abc import Callable, Collection

from hideout.config import get_config
from hideout.ollama import chat
from hideout.search import Hit, format_hit, rrf


def prompt_messages(question: str, hits: list[Hit]) -> list[dict]:
    blocks = []
    for i, hit in enumerate(hits, start=1):
        c = hit.chunk
        section = " > ".join([c.chapter, *c.headings])
        page = f"PDF page {c.page}" if c.page else "unknown page"
        if c.printed is not None:
            page += f", printed {c.printed}"
        blocks.append(
            f"[{i}] {c.book} | {c.file} | {section} | {page}\n{c.text}"
        )
    context = "\n\n----\n\n".join(blocks)
    return [
        {"role": "system", "content": get_config().system_prompt},
        {
            "role": "user",
            "content": f"Excerpts:\n\n{context}\n\nQuestion: {question}",
        },
    ]


def ask(
    question: str,
    k: int = 8,
    on_token: Callable[[str], None] | None = None,
    books: Collection[str] | None = None,
) -> tuple[str, list[Hit]]:
    hits = rrf(question, k=k, books=books)
    if not hits:
        msg = "No matching excerpts in the index."
        if on_token:
            on_token(msg)
        return msg, []
    parts: list[str] = []
    for piece in chat(prompt_messages(question, hits), stream=True):
        parts.append(piece)
        if on_token:
            on_token(piece)
    return "".join(parts).strip(), hits


def print_hits(hits: list[Hit]) -> None:
    for i, hit in enumerate(hits, start=1):
        print(f"--- {i} ---")
        print(format_hit(hit))
        print()

File: src/ask.ts
import { getConfig } from "./config.js";
import { chat } from "./llm.js";
import { formatHit, type Hit, rrf } from "./search.js";

export function promptMessages(
  question: string,
  hits: Hit[],
): { role: string; content: string }[] {
  const blocks: string[] = [];
  hits.forEach((hit, i) => {
    const c = hit.chunk;
    const section = [c.chapter, ...c.headings].join(" > ");
    const parts = [c.book, c.file, section];
    if (c.page) {
      let page = `PDF page ${c.page}`;
      if (c.printed != null) page += `, printed ${c.printed}`;
      parts.push(page);
    }
    blocks.push(`[${i + 1}] ${parts.join(" | ")}\n${c.text}`);
  });
  const context = blocks.join("\n\n----\n\n");
  return [
    { role: "system", content: getConfig().systemPrompt },
    { role: "user", content: `Excerpts:\n\n${context}\n\nQuestion: ${question}` },
  ];
}

export async function ask(
  question: string,
  opts: {
    k?: number;
    onToken?: (token: string) => void;
    books?: Iterable<string> | null;
  } = {},
): Promise<[string, Hit[]]> {
  const hits = await rrf(question, { k: opts.k ?? 8, books: opts.books });
  if (!hits.length) {
    const msg = "No matching excerpts in the index.";
    opts.onToken?.(msg);
    return [msg, []];
  }
  const parts: string[] = [];
  for await (const piece of chat(promptMessages(question, hits), true)) {
    parts.push(piece);
    opts.onToken?.(piece);
  }
  return [parts.join("").trim(), hits];
}

export function printHits(hits: Hit[]): void {
  hits.forEach((hit, i) => {
    console.log(`--- ${i + 1} ---`);
    console.log(formatHit(hit));
    console.log();
  });
}
